### PR TITLE
[tests-only] first try of a contract test

### PR DESCRIPTION
**Pull request by individual-it**
A first POC of a simple contract test with pact.io

to run:

- consumer: `yarn test:pact`
- provider: `make test` or `go test  -v ./tests/pact`

### CODE REVIEW

File: tests/pact/index.pact.spec.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    const greetRequest = {
+      uponReceiving: 'a greeting',
+      withRequest: {
+        method: 'POST',
+        path: '/api/v0/greet',
+        body: REQUEST_DATA
+      }
+    }
+
+    beforeEach(() => {
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:32]: Codacy found an issue: ['beforeEach' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      return provider.addInteraction(interaction)
+    })
+
+    // add expectations
+    it('returns a successful body', () => {
+      return Hello_Greet({
+        $domain: provider.mockService.baseUrl,
+        body: REQUEST_DATA
+      }).then(response => {
+        expect(response.data).toEqual(RESPONSE_DATA)
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:47]: Codacy found an issue: ['expect' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+'use strict'
+
+const { pactWith } = require('jest-pact')
+const { Hello_Greet } = require('../../ui/client/hello/index')
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:4]: Codacy found an issue: [Identifier 'Hello_Greet' is not in camel case.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      const interaction = {
+        state: 'i am greeted with my name',
+        ...greetRequest,
+        willRespondWith: successResponse
+      }
+      return provider.addInteraction(interaction)
+    })
+
+    // add expectations
+    it('returns a successful body', () => {
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:42]: Codacy found an issue: ['it' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
+'use strict'
+
+const { pactWith } = require('jest-pact')
+const { Hello_Greet } = require('../../ui/client/hello/index')
+
+pactWith({ consumer: 'OCIS-Hello_UI', provider: 'OCIS-Hello_API' }, provider => {
+  describe('Hello API', () => {
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:7]: Codacy found an issue: ['describe' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js (hunk 1 above)
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:32]: Codacy found an issue: ['beforeEach' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js (hunk 5 above)
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:7]: Codacy found an issue: ['describe' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js (hunk 3 above)
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:4]: Codacy found an issue: [Identifier 'Hello_Greet' is not in camel case.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js (hunk 2 above)
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:47]: Codacy found an issue: ['expect' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: tests/pact/index.pact.spec.js (hunk 4 above)
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:42]: Codacy found an issue: ['it' is not defined.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

**Comment by individual-it** [comment]
tests run successfully in CI but yarn fails on ARM
image

File: tests/pact/index.pact.spec.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    it('returns an error with a wrong field', async () => {
+      const interaction = {
+        state: 'i have sent a wrong field name',
+        ...greetRequestWithInvalidField,
+        willRespondWith: preconditionFailedResponse
+      }
+      await provider.addInteraction(interaction)
+
+      try {
```
Comment (ownclouders) [inline tests/pact/index.pact.spec.js:82]: Codacy found an issue: [Unnecessary block.](https://app.codacy.com/gh/owncloud/ocis-hello/pullRequest?prid=5869027)

File: .drone.star [hunk 7]
```diff
@@ -119,6 +122,10 @@ def testing(ctx):
         'image': 'webhippie/golang:1.13',
         'pull': 'always',
         'commands': [
+          'apk add --no-cache --virtual build-dependencies build-base',
+          'apk --no-cache add ca-certificates wget bash && wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub && wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.29-r0/glibc-2.29-r0.apk && apk add glibc-2.29-r0.apk',
+          'curl -fsSL https://raw.githubusercontent.com/pact-foundation/pact-ruby-standalone/master/install.sh | bash',
```
Comment (individual-it) [inline .drone.star:127]: could be moved in a separate docker container

File: .drone.star [hunk 8]
```diff
@@ -38,10 +38,13 @@ def testing(ctx):
         'image': 'webhippie/nodejs:latest',
         'pull': 'always',
         'commands': [
+          'apk add --no-cache --virtual build-dependencies build-base',
+          'apk --no-cache add ca-certificates wget bash && wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub && wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.29-r0/glibc-2.29-r0.apk && apk add glibc-2.29-r0.apk',
```
Comment (individual-it) [inline .drone.star:42]: could be moved in a separate docker container

File: .drone.star [hunk 9]
```diff
@@ -133,6 +140,7 @@ def testing(ctx):
         'image': 'plugins/codacy:1',
         'pull': 'always',
         'settings': {
+          'pattern': '/drone/src/*.out',
```
Comment (individual-it) [inline .drone.star:143]: the new node modules contain some .out files that confuses codacy

File: package.json [hunk 10]
```diff
@@ -17,10 +17,16 @@
     "watch": "rollup -c -w",
     "test": "echo 'Not implemented'",
     "generate-api": "node node_modules/swagger-vue-generator/bin/generate-api.js --package-version v0 --source pkg/proto/v0/hello.swagger.json --moduleName hello --destination ui/client/hello/index.js",
+    "test:pact": "jest tests/pact --runInBand --env=node",
     "acceptance-tests": "cucumber-js --require-module @babel/register --require-module @babel/polyfill --require ${TEST_INFRA_DIRECTORY}/acceptance/setup.js --require ui/tests/acceptance/stepDefinitions --require ${TEST_INFRA_DIRECTORY}/acceptance/stepDefinitions --format node_modules/cucumber-pretty -t \"${TEST_TAGS:-not @skip and not @skipOnOC10}\""
   },
+  "optionalDependencies": {
```
Comment (individual-it) [inline package.json:23]: pact is not compatible with the arm platform, so the dependencies need to be optional

**Comment by individual-it** [comment]
works now, ready for review and comments

File: package.json [hunk 11]
```diff
@@ -17,10 +17,16 @@
...
+    "@pact-foundation/pact": "^9.11.0",
+    "@pact-foundation/pact-node": "^10.9.7",
+    "jest-pact": "^0.6.0"
+  },
   "devDependencies": {
-    "@babel/core": "^7.7.7",
+    "@babel/core": "^.7.7",
```
Comment (phil-davis) [inline package.json:29]: Is this intentional?

Comment (individual-it) [reply package.json:29]: yarn has done it

**Comment by individual-it** [comment]
rebased on current master

**Comment by individual-it** [comment]
we are using this system now in the SDK, if needed between the micro-services we could come back again. closing in the meantime

**Pull request closed without merging**